esphome/components/nextion/nextion_upload_arduino.cpp

```diff
@@ -203,135 +237,118 @@ bool Nextion::upload_tft() {
+  ESP_LOGD(TAG, "  URL: %s", this->tft_url_.c_str());
+  ESP_LOGD(TAG, "  File size: %d bytes", this->content_length_);
+  ESP_LOGD(TAG, "  Free heap: %" PRIu32, this->get_free_heap_());
 
-  if (this->transfer_buffer_ == nullptr) {
-    ExternalRAMAllocator<uint8_t> allocator(ExternalRAMAllocator<uint8_t>::ALLOW_FAILURE);
-    // NOLINTNEXTLINE(readability-static-accessed-through-instance)
-    ESP_LOGD(TAG, "Allocating buffer size %d, Heap size is %u", chunk_size, ESP.getFreeHeap());
-    this->transfer_buffer_ = allocator.allocate(chunk_size);
-    if (this->transfer_buffer_ == nullptr) {  // Try a smaller size
-      ESP_LOGD(TAG, "Could not allocate buffer size: %d trying 4096 instead", chunk_size);
-      chunk_size = 4096;
-      ESP_LOGD(TAG, "Allocating %d buffer", chunk_size);
-      this->transfer_buffer_ = allocator.allocate(chunk_size);
-
-      if (!this->transfer_buffer_)
-        return this->upload_end_(false);
-    }
+  // Proceed with the content download as before
 
-    this->transfer_buffer_size_ = chunk_size;
-  }
-
-  // NOLINTNEXTLINE(readability-static-accessed-through-instance)
-  ESP_LOGD(TAG, "Updating tft from \"%s\" with a file size of %d using %zu chunksize, Heap Size %d",
-           this->tft_url_.c_str(), this->content_length_, this->transfer_buffer_size_, ESP.getFreeHeap());
+  ESP_LOGV(TAG, "Starting transfer by chunks loop");
 
-  int result = 0;
+  uint32_t position = 0;
   while (this->content_length_ > 0) {
-    result = this->upload_by_chunks_(&http, result);
-    if (result < 0) {
-      ESP_LOGD(TAG, "Error updating Nextion!");
+    int upload_result = upload_by_chunks_(http_client, position);
+    if (upload_result < 0) {
+      ESP_LOGE(TAG, "Error uploading TFT to Nextion!");
+      ESP_LOGD(TAG, "Close HTTP connection");
+      http_client.end();
+      ESP_LOGV(TAG, "Connection closed");
       return this->upload_end_(false);
     }
     App.feed_wdt();
-    // NOLINTNEXTLINE(readability-static-accessed-through-instance)
-    ESP_LOGD(TAG, "Heap Size %d, Bytes left %d", ESP.getFreeHeap(), this->content_length_);
+    ESP_LOGV(TAG, "Free heap: %" PRIu32 ", Bytes left: %" PRIu32, this->get_free_heap_(), this->content_length_);
   }
-  ESP_LOGD(TAG, "Successfully updated Nextion!");
 
-  return this->upload_end_(true);
+  ESP_LOGD(TAG, "Successfully uploaded TFT to Nextion!");
+
+  ESP_LOGD(TAG, "Close HTTP connection");
+  http_client.end();
+  ESP_LOGV(TAG, "Connection closed");
+  return upload_end_(true);
 }
 
 bool Nextion::upload_end_(bool successful) {
+  ESP_LOGD(TAG, "Nextion TFT upload finished: %s", YESNO(successful));
   this->is_updating_ = false;
-  ESP_LOGD(TAG, "Restarting Nextion");
-  this->soft_reset();
+  this->ignore_is_setup_ = false;
+
   if (successful) {
-    delay(1500);  // NOLINT
-    ESP_LOGD(TAG, "Restarting esphome");
+    ESP_LOGD(TAG, "Restarting ESPHome");
+    delay(1500);    // NOLINT
     ESP.restart();  // NOLINT(readability-static-accessed-through-instance)
```

Comment:
This is abstracted out in "esphome/core/hal.h"
```suggestion
    arch_restart();
```